### ISSUE TITLE
Neovim locks up on substitute when set inccommand=nosplit and set gdefault

### ISSUE THREAD

**Issue by glacambre** [neovim/neovim#7244]
- `nvim --version`: NVIM v0.2.1-855-gcb912a3ed
- Operating system/version: Linux 4.12.5-gentoo
- Terminal name/version: rxvt-unicode (urxvt) v9.21
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`
```
echo '{' | nvim -u NORC -c "set inccommand=nosplit | set gdefault" -
:s/{\n
```

### Actual behaviour
Neovim is unresponsive and eats 100% of one of my CPU cores until C-c is sent.

### Expected behaviour
No lock up.

**Comment by KillTheMule** [comment]
Can confirm, also happens with `inccommand=split`.

I'm reworking `inccomand` substitution in https://github.com/neovim/neovim/pull/7231, but it also happens with that PR, just at a different point (you get to actually type `:s/` only). I'll check it out in more depth over the weekend, thanks for the report and the reproducible example :)

**Comment by KillTheMule** [comment]
Could you try https://github.com/neovim/neovim/pull/7249? Fixes it for me :)

**Comment by KillTheMule** [comment]
This is multiline territory, so I'm working on it in https://github.com/neovim/neovim/pull/7231. I've found some incorrect code which fixes 
```
echo -ne '{\n\n{' | build/bin/nvim -u NORC -c "set inccommand=nosplit | set gdefault" -
:%s/
```

but seems not to fix
```
echo '{' | nvim -u NORC -c "set inccommand=nosplit | set gdefault" -
:s/{\n
```

I mostly believe the code in `do_sub` is correct (minus highlighting), and the backtraces seem to support that, but it's getting to complicated for me. I've interrupted the locked nvim instance from gdb randomly and recorded the backtraces in https://gist.github.com/KillTheMule/474c1d5ac6b35b3b6fb07a5f9bb8cdc7. @justinmk @bfredl @jamessan  could someone have a look for diagnostics? This is with PR #7231, but as far as I can see, almost no code is run from `do_sub` anyways. Looks really weird to me, `do_sub` is called with `eap == NULL`, but there's no SEGFAULT, but `ascii_iswhite` is called with a very strange number... Any help appreciated!
